File: src/main/java/me/oriharel/machinery/data/ZenCoinChance.java
package me.oriharel.machinery.data;

import me.oriharel.machinery.machine.MachineResourceGetProcess;

import java.util.Objects;

public class ZenCoinChance implements ChanceableOperation<Integer, MachineResourceGetProcess> {

    private Range range;

    public ZenCoinChance(Range range) {
        this.range = range;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ZenCoinChance that = (ZenCoinChance) o;
        return Objects.equals(range, that.range);
    }

    @Override
    public int hashCode() {
        return Objects.hash(range);
    }

    @Override
    public String toString() {
        return "ZenCoinChance{" +
                "range=" + range +
                '}';
    }

    @Override
    public Integer getChanced() {
        return range.random();
    }

    @Override
    public void executeChanceOperation(MachineResourceGetProcess machineResourceGetProcess) {
        machineResourceGetProcess.addZenCoinsGained(getChanced());
    }
}

File: src/main/java/me/oriharel/machinery/data/ZenCoinChance.kt
package me.oriharel.machinery.data

import me.oriharel.machinery.machine.MachineResourceGetProcess
import java.util.*

class ZenCoinChance(private val range: Range) : ChanceableOperation<Int?, MachineResourceGetProcess> {
    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val that = o as ZenCoinChance
        return range == that.range
    }

    override fun hashCode(): Int {
        return Objects.hash(range)
    }

    override fun toString(): String {
        return "ZenCoinChance{" +
                "range=" + range +
                '}'
    }

    override fun getChanced(lootModifier: Double): Int? {
        return (range.random() * lootModifier).toInt()
    }

    override fun executeChanceOperation(machineResourceGetProcess: MachineResourceGetProcess, lootModifier: Double) {
        machineResourceGetProcess.addZenCoinsGained(getChanced(lootModifier)!!.toLong())
    }

}